File: cpp/controllers_py/joint_velocity_filter_test.py
# Copyright 2020 DeepMind Technologies Limited.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tests for joint_velocity_filter PyBind11 module."""

from absl.testing import absltest
from absl.testing import parameterized
from dm_control import mujoco
from dm_control.suite import humanoid
from dm_robotics.controllers import joint_velocity_filter
import numpy as np

_MjGeom = mujoco.wrapper.mjbindings.enums.mjtObj.mjOBJ_GEOM
_MjBody = mujoco.wrapper.mjbindings.enums.mjtObj.mjOBJ_BODY
_MjSite = mujoco.wrapper.mjbindings.enums.mjtObj.mjOBJ_SITE
mjlib = mujoco.wrapper.mjbindings.mjlib


def _set_joint_velocities(physics, joint_ids, joint_velocities):
  """Sets the joint velocities in physics for a subset of joints."""
  for i in range(0, physics.model.nv):
    physics.data.qvel[i] = 0.0
  for joint_id, velocity in zip(sorted(joint_ids), joint_velocities):
    dof_adr = physics.model.jnt_dofadr[joint_id]
    physics.data.qvel[dof_adr] = velocity


class JointVelocityFilterTest(absltest.TestCase):

  def test_parameters_attributes(self):
    params = joint_velocity_filter.Parameters()

    attributes = sorted(
        [attr for attr in dir(params) if not attr.startswith("_")]
    )
    expected_attributes = sorted([
        "model",
        "joint_ids",
        "integration_timestep",
        "enable_joint_position_limits",
        "joint_position_limit_velocity_scale",
        "minimum_distance_from_joint_position_limit",
        "enable_joint_velocity_limits",
        "joint_velocity_magnitude_limits",
        "enable_collision_avoidance",
        "use_minimum_distance_contacts_only",
        "collision_avoidance_normal_velocity_scale",
        "minimum_distance_from_collisions",
        "collision_detection_distance",
        "collision_pairs",
        "check_solution_validity",
        "max_qp_solver_iterations",
        "solution_tolerance",
        "use_adaptive_step_size",
        "log_collision_warnings",
    ])
    self.assertEqual(expected_attributes, attributes)


@parameterized.named_parameters(
    ("use_adaptive_step_size", True),
    ("do_not_use_adaptive_step_size", False),
)
class JointVelocityFilterParameterizedTest(absltest.TestCase):

  def test_filter_attributes(self, use_adaptive_step_size):
    physics = humanoid.Physics.from_xml_string(*humanoid.get_model_and_assets())

    params = joint_velocity_filter.Parameters()
    params.model = physics.model
    params.joint_ids = [19, 20, 21]
    params.integration_timestep = 1.0
    params.use_adaptive_step_size = use_adaptive_step_size
    joint_vel_filter = joint_velocity_filter.JointVelocityFilter(params)

    self.assertTrue(hasattr(joint_vel_filter, "filter_joint_velocities"))

  def test_solution_realizes_target(self, use_adaptive_step_size):
    physics = humanoid.Physics.from_xml_string(*humanoid.get_model_and_assets())

    params = joint_velocity_filter.Parameters()
    params.model = physics.model
    params.joint_ids = [16, 17, 18]
    params.integration_timestep = 1.0
    params.solution_tolerance = 1.0e-15
    params.max_qp_solver_iterations = 300
    params.use_adaptive_step_size = use_adaptive_step_size
    joint_vel_filter = joint_velocity_filter.JointVelocityFilter(params)

    # Set target to a realizable velocity and solve.
    target_velocity = [
        0.0450566,
        0.0199436,
        0.0199436,
    ]
    solution = joint_vel_filter.filter_joint_velocities(
        physics.data, target_velocity
    )
    _set_joint_velocities(physics, params.joint_ids, solution)

    # Realized joint velocity must be within the specified tolerance of the
    # target velocity.
    diff_norm = np.linalg.norm(solution - np.array(target_velocity))
    self.assertLess(diff_norm, params.solution_tolerance)

  def test_solution_with_all_constraints_not_in_collision(
      self, use_adaptive_step_size
  ):
    physics = humanoid.Physics.from_xml_string(*humanoid.get_model_and_assets())

    # Increase collision detection margin for all geoms.
    for i in range(0, physics.model.ngeom):
      physics.model.geom_margin[i] = 0.01

    # Place the humanoid in a position where the left hand can collide with the
    # floor if it moves down.
    physics.data.qpos[2] = 0.3
    mjlib.mj_fwdPosition(physics.model.ptr, physics.data.ptr)

    # Set params with collision avoidance and full list of constraints.
    params = joint_velocity_filter.Parameters()
    params.model = physics.model
    params.joint_ids = [19, 20, 21]
    params.integration_timestep = 0.005  # 5ms

    params.enable_joint_position_limits = True
    params.joint_position_limit_velocity_scale = 0.95
    params.minimum_distance_from_joint_position_limit = 0.01  # ~0.5deg.

    params.enable_joint_velocity_limits = True
    params.joint_velocity_magnitude_limits = [0.5, 0.5, 0.5]

    params.enable_collision_avoidance = True
    params.collision_avoidance_normal_velocity_scale = 0.01
    params.minimum_distance_from_collisions = 0.005
    params.collision_detection_distance = 10.0
    params.collision_pairs = [(
        ["left_upper_arm", "left_lower_arm", "left_hand"],
        ["floor"],
    )]

    params.check_solution_validity = True
    params.solution_tolerance = 1e-6
    params.max_qp_solver_iterations = 300
    params.use_adaptive_step_size = use_adaptive_step_size
    joint_vel_filter = joint_velocity_filter.JointVelocityFilter(params)

    # Approximate the distance of the left hand and floor geoms by the
    # difference in Z components minus the radius.
    lhand_radius = physics.named.model.geom_size["left_hand"][0]
    lhand_floor_dist = (
        physics.named.data.geom_xpos["left_hand"][2]
        - physics.named.data.geom_xpos["floor"][2]
        - lhand_radius
    )

    # We make the left hand move down towards the plane.
    # These values were chosen to produce an initial downward motion that later
    # isn't purely downwards but would still hit the floor.
    target_velocity = [-0.02, 0.02, -0.02]

    # Compute velocities and integrate, for 5000 steps.
    for _ in range(0, 5000):
      # Compute joint velocities.
      solution = joint_vel_filter.filter_joint_velocities(
          physics.data, target_velocity
      )

      # Set joint velocities, integrate, and run MuJoCo kinematics.
      _set_joint_velocities(physics, params.joint_ids, solution)
      mjlib.mj_integratePos(
          physics.model.ptr,
          physics.data.qpos,
          physics.data.qvel,
          params.integration_timestep.total_seconds(),
      )
      mjlib.mj_fwdPosition(physics.model.ptr, physics.data.ptr)

      # Compute the new distance between the floor and the left hand.
      # We expect the left hand site to get closer to the floor and settle at
      # around <0.006.
      new_lhand_floor_dist = (
          physics.named.data.geom_xpos["left_hand"][2]
          - physics.named.data.geom_xpos["floor"][2]
          - lhand_radius
      )
      self.assertLess(new_lhand_floor_dist, max(0.006, lhand_floor_dist))
      lhand_floor_dist = new_lhand_floor_dist

      # Ensure there is no contact between any left arm geom and the floor.
      for contact in physics.data.contact:
        geom1_name = physics.model.id2name(contact.geom1, _MjGeom)
        geom2_name = physics.model.id2name(contact.geom2, _MjGeom)
        if contact.dist < params.minimum_distance_from_collisions:
          is_any_left_hand = (
              geom1_name == "left_hand" or geom2_name == "left_hand"
          )
          is_any_left_upperarm = (
              geom1_name == "left_upper_arm" or geom2_name == "left_upper_arm"
          )
          is_any_left_lowerarm = (
              geom1_name == "left_lower_arm" or geom2_name == "left_lower_arm"
          )
          is_any_left_arm = (
              is_any_left_hand or is_any_left_upperarm or is_any_left_lowerarm
          )
          is_any_floor = geom1_name == "floor" or geom2_name == "floor"

          self.assertFalse(is_any_left_arm and is_any_floor)

  def test_invalid_parameters_throws(self, use_adaptive_step_size):
    physics = humanoid.Physics.from_xml_string(*humanoid.get_model_and_assets())

    params = joint_velocity_filter.Parameters()
    params.model = physics.model
    params.joint_ids = [19, 20, 21]
    params.integration_timestep = -1.0  # Not valid.
    params.use_adaptive_step_size = use_adaptive_step_size
    with self.assertRaises(Exception):
      _ = joint_velocity_filter.JointVelocityFilter(params)


if __name__ == "__main__":
  absltest.main()

File: cpp/controllers/lsqp/tests/joint_velocity_filter_test.cc
// Copyright 2022 DeepMind Technologies Limited
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "dm_robotics/controllers/lsqp/joint_velocity_filter.h"
#include <algorithm>
#include <array>

#include "dm_robotics/support/status-matchers.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/container/btree_set.h"
#include "absl/time/time.h"
#include "absl/types/span.h"
#include "dm_robotics/controllers/lsqp/test_utils.h"
#include "dm_robotics/mujoco/defs.h"
#include "dm_robotics/mujoco/test_with_mujoco_model.h"
#include "dm_robotics/mujoco/types.h"
#include "dm_robotics/mujoco/utils.h"
#include "Eigen/Core"
#include <mujoco/mujoco.h>  //NOLINT

namespace dm_robotics {
namespace {

using ::dm_robotics::testing::SetSubsetOfJointVelocities;
using ::dm_robotics::testing::TestWithMujocoModel;
using ::testing::ValuesIn;
using ::testing::WithParamInterface;

class JointVelocityFilterTest :
    public TestWithMujocoModel,
    public WithParamInterface<bool> {};

constexpr bool kJointVelocityFilterParameterSet[] = {
    true, false
};

INSTANTIATE_TEST_SUITE_P(
    JointVelocityFilterTests,
    JointVelocityFilterTest,
    ValuesIn(kJointVelocityFilterParameterSet));

TEST_P(JointVelocityFilterTest,
       SolutionIsOkAndRealizesTarget) {
  LoadModelFromXmlPath(kDmControlSuiteHumanoidXmlPath);
  const std::array<double, 3> kTargetVelocity = {
      0.0450566, 0.0199436, 0.0199436
  };
  const absl::btree_set<int> kJointIds{16, 17, 18};
  constexpr double kSolutionTolerance = 1.0e-15;

  // Instantiate filter and solve once.
  JointVelocityFilter::Parameters params;
  params.model = model_.get();
  params.joint_ids = kJointIds;
  params.integration_timestep = absl::Seconds(1);
  params.solution_tolerance = kSolutionTolerance;
  params.use_adaptive_step_size = GetParam();

  ASSERT_OK(JointVelocityFilter::ValidateParameters(params));
  JointVelocityFilter filter(params);
  ASSERT_OK_AND_ASSIGN(absl::Span<const double> solution,
                       filter.FilterJointVelocities(*data_, kTargetVelocity));
  // Ensure the computed joint velocities are within the tolerance of the target
  // joint velocities.
  auto solution_vec = Eigen::Map<const Eigen::VectorXd>(
          solution.data(), solution.size());
  auto target_vec = Eigen::Map<const Eigen::VectorXd>(
          kTargetVelocity.data(), kTargetVelocity.size());
  double diff_norm = (solution_vec - target_vec).lpNorm<Eigen::Infinity>();
  EXPECT_LE(diff_norm, kSolutionTolerance);
}

TEST_P(JointVelocityFilterTest,
       SolutionWithAllConstraintsAndNotInCollision) {
  bool use_adaptive_step_size = GetParam();
  LoadModelFromXmlPath(kDmControlSuiteHumanoidXmlPath);

  // Place the humanoid in a position where the left hand can collide with the
  // floor if it moves down.
  data_->qpos[2] = 0.3;
  mj_fwdPosition(model_.get(), data_.get());

  // We make the left hand move down towards the plane.
  // These values were chosen to produce an initial downward motion that later
  // isn't purely downwards but would still hit the floor.
  const std::array<double, 3> kTargetVelocity = {-0.02, 0.02, -0.02};
  const absl::btree_set<int> kJointIds{19, 20, 21};
  const double kSolutionTolerance = use_adaptive_step_size ? 1.0e-3 : 1.0e-6;

  JointVelocityFilter::Parameters params;
  params.model = model_.get();
  params.joint_ids = kJointIds;
  params.integration_timestep = absl::Milliseconds(5);

  params.enable_joint_position_limits = true;
  params.joint_position_limit_velocity_scale = 0.95;
  params.minimum_distance_from_joint_position_limit = 0.01;  // ~0.5deg.

  params.enable_joint_velocity_limits = true;
  params.joint_velocity_magnitude_limits = {0.5, 0.5, 0.5};

  params.enable_collision_avoidance = true;
  params.collision_avoidance_normal_velocity_scale = 0.01;
  params.minimum_distance_from_collisions = 0.005;
  params.collision_detection_distance = 10.0;
  params.collision_pairs = {
      CollisionPair(GeomGroup{"left_upper_arm", "left_lower_arm", "left_hand"},
                    GeomGroup{"floor"})};

  params.check_solution_validity = true;
  params.max_qp_solver_iterations = 300;
  params.solution_tolerance = kSolutionTolerance;
  params.use_adaptive_step_size = use_adaptive_step_size;

  ASSERT_OK(JointVelocityFilter::ValidateParameters(params));
  JointVelocityFilter filter(params);

  // Convert to geom pairs and get geom IDs to query MuJoCo for collision
  // information.
  auto geom_pairs = CollisionPairsToGeomIdPairs(
      *model_, params.collision_pairs, false, false);
  int left_hand_id =
      mj_name2id(model_.get(), mjtObj::mjOBJ_GEOM, "left_hand");
  int floor_id = mj_name2id(model_.get(), mjtObj::mjOBJ_GEOM, "floor");
  auto maybe_dist = ComputeMinimumContactDistance(*model_, *data_,
                                                  left_hand_id, floor_id, 10.0);
  ASSERT_TRUE(maybe_dist.has_value());
  double left_hand_to_floor_dist = *maybe_dist;

  // Compute velocities and integrate, for 5000 steps.
  // We expect that the distance from the hand to the floor will decrease from
  // iteration to iteration, and eventually settle at <0.006. None of the geoms
  // in the collision pair should ever penetrate.
  for (int i = 0; i < 5000; ++i) {
    ASSERT_OK_AND_ASSIGN(
        absl::Span<const double> solution,
        filter.FilterJointVelocities(*data_, kTargetVelocity));
    SetSubsetOfJointVelocities(*model_, kJointIds, solution, data_.get());
    mj_integratePos(model_.get(), data_->qpos, data_->qvel,
                            absl::ToDoubleSeconds(params.integration_timestep));
    mj_fwdPosition(model_.get(), data_.get());

    // Ensure the new distance always decreases from iteration to iteration,
    // until it settles.
    auto maybe_new_dist = ComputeMinimumContactDistance(
        *model_, *data_, left_hand_id, floor_id, 10.0);
    ASSERT_TRUE(maybe_new_dist.has_value());
    EXPECT_LE(*maybe_new_dist, std::max(0.006, left_hand_to_floor_dist));
    left_hand_to_floor_dist = *maybe_new_dist;

    // Ensure no contacts are detected on the collision pairs.
    ASSERT_OK_AND_ASSIGN(
        int num_contacts,
        ComputeContactsForGeomPairs(*model_, *data_, geom_pairs,
                                    params.minimum_distance_from_collisions,
                                    absl::Span<mjContact>()));
    EXPECT_EQ(num_contacts, 0);
  }
}

TEST_P(
    JointVelocityFilterTest,
    SolutionWithAllConstraintsAndNullspaceIsOkAndNotInCollisionSingleContacts) {
  bool use_adaptive_step_size = GetParam();
  LoadModelFromXmlPath(kDmControlSuiteHumanoidXmlPath);

  // Place the humanoid in a position where the left hand can collide with the
  // floor if it moves down.
  data_->qpos[2] = 0.3;
  mj_fwdPosition(model_.get(), data_.get());

  // We make the left hand move down towards the plane.
  // These values were chosen to produce an initial downward motion that later
  // isn't purely downwards but would still hit the floor.
  const std::array<double, 3> kTargetVelocity = {-0.02, 0.02, -0.02};
  const absl::btree_set<int> kJointIds{19, 20, 21};
  const double kSolutionTolerance = use_adaptive_step_size ? 1.0e-3 : 1.0e-6;

  JointVelocityFilter::Parameters params;
  params.model = model_.get();
  params.joint_ids = kJointIds;
  params.integration_timestep = absl::Milliseconds(5);

  params.enable_joint_position_limits = true;
  params.joint_position_limit_velocity_scale = 0.95;
  params.minimum_distance_from_joint_position_limit = 0.01;  // ~0.5deg.

  params.enable_joint_velocity_limits = true;
  params.joint_velocity_magnitude_limits = {0.5, 0.5, 0.5};

  params.enable_collision_avoidance = true;
  params.use_minimum_distance_contacts_only = true;
  params.collision_avoidance_normal_velocity_scale = 0.01;
  params.minimum_distance_from_collisions = 0.005;
  params.collision_detection_distance = 10.0;
  params.collision_pairs = {
      CollisionPair(GeomGroup{"left_upper_arm", "left_lower_arm", "left_hand"},
                    GeomGroup{"floor"})};

  params.check_solution_validity = true;
  params.solution_tolerance = kSolutionTolerance;
  params.use_adaptive_step_size = use_adaptive_step_size;

  ASSERT_OK(JointVelocityFilter::ValidateParameters(params));
  JointVelocityFilter filter(params);

  // Convert to geom pairs and get geom IDs to query MuJoCo for collision
  // information.
  auto geom_pairs = CollisionPairsToGeomIdPairs(
      *model_, params.collision_pairs, false, false);
  int left_hand_id =
      mj_name2id(model_.get(), mjtObj::mjOBJ_GEOM, "left_hand");
  int floor_id = mj_name2id(model_.get(), mjtObj::mjOBJ_GEOM, "floor");
  auto maybe_dist = ComputeMinimumContactDistance(*model_, *data_,
                                                  left_hand_id, floor_id, 10.0);
  ASSERT_TRUE(maybe_dist.has_value());
  double left_hand_to_floor_dist = *maybe_dist;

  // Compute velocities and integrate, for 5000 steps.
  // We expect that the distance from the hand to the floor will decrease from
  // iteration to iteration, and eventually settle at <0.006. None of the geoms
  // in the collision pair should ever penetrate.
  for (int i = 0; i < 5000; ++i) {
    ASSERT_OK_AND_ASSIGN(
        absl::Span<const double> solution,
        filter.FilterJointVelocities(*data_, kTargetVelocity));
    SetSubsetOfJointVelocities(*model_, kJointIds, solution, data_.get());
    mj_integratePos(model_.get(), data_->qpos, data_->qvel,
                            absl::ToDoubleSeconds(params.integration_timestep));
    mj_fwdPosition(model_.get(), data_.get());

    // Ensure the new distance always decreases from iteration to iteration,
    // until it settles.
    auto maybe_new_dist = ComputeMinimumContactDistance(
        *model_, *data_, left_hand_id, floor_id, 10.0);
    ASSERT_TRUE(maybe_new_dist.has_value());
    EXPECT_LE(*maybe_new_dist, std::max(0.006, left_hand_to_floor_dist));
    left_hand_to_floor_dist = *maybe_new_dist;

    // Ensure no contacts are detected on the collision pairs.
    ASSERT_OK_AND_ASSIGN(
        int num_contacts,
        ComputeContactsForGeomPairs(*model_, *data_, geom_pairs,
                                    params.minimum_distance_from_collisions,
                                    absl::Span<mjContact>()));
    EXPECT_EQ(num_contacts, 0);
  }
}

}  // namespace
}  // namespace dm_robotics
